feddlib/core/AceFemAssembly/specific/AssembleFEGeneralizedNewtonian_def.hpp

```diff
@@ -640,8 +663,246 @@ namespace FEDD
+        // via the inverse mapping xi = Binv(x-tau) where tau is p0
+        // tau = [this->nodesRefConfig_.at(0).at(0) ;  this->nodesRefConfig_.at(0).at(1)]
+
+        for (int l = 0; l < QuadPtsPhy.size(); l++)
+        {
+            for (int p = 0; p < dim; p++)
+            {
+                for (int q = 0; q < dim; q++)
+                {
+                    QuadPtsMapped[l][p]  += Binv[p][q] * ( QuadPtsPhy[l][q] - this->nodesRefConfig_.at(0).at(q));
+                }
+            }
+        }
+
+        // Now we have successfully mapped the global quadrature points onto one of the reference faces (2D line / 3D surface)
+        // But we still have to consider the case that if the global surface was mapped onto the diagonal line/ diagonal surface we
+        // have to scale the weights and quadrature points by the factor area change so in 2D it is just the length so sqrt(2) [0 sqrt(2)] and in 3D it should be the area which is 0.866025403784439
+        
+        // Check if a quadrature point lies on the diagonal line
+        // We consider the point to be on the diagonal if both its x and y components are nonzero
+        // https://www.math.ntnu.no/emner/TMA4130/2021h/lectures/CompositeQuadrature.pdf
+        if (dim==2)
+        {
+            if (  (std::fabs( QuadPtsMapped[0][0] - 0.0) >  std::numeric_limits<double>::epsilon()  )   &&  ( (QuadPtsMapped[0][1]  - 0.0) >  std::numeric_limits<double>::epsilon()    ) ) // if the x and y component of quadrature point are non-zero we are on the diagonal but also only if the quadrature point was not defined in corners of element
+            {
+                for (int l = 0; l < QuadPtsPhy.size(); l++)
+                {
+                QuadWeights[l] = std::sqrt(2.0) * QuadWeights[l];
+                    for (int q = 0; q < dim; q++)
+                    {
+                    QuadPtsMapped[l][q] = std::sqrt(2.0) * QuadPtsMapped[l][q];
+                    }
+                }
+            }
+        }
+        else if (dim==3) // This has to be checked again if we really just multiply the surface are of the triangle?
+        {
+            if (  (std::fabs( QuadPtsMapped[0][0] - 0.0) >  std::numeric_limits<double>::epsilon()  )   &&  ( (QuadPtsMapped[0][1]  - 0.0) >  std::numeric_limits<double>::epsilon()    ) &&  ( (QuadPtsMapped[0][2]  - 0.0) >  std::numeric_limits<double>::epsilon()    ) ) // if the x and y component of quadrature point are non-zero we are on the diagonal but also only if the quadrature point was not defined in corners of element
+            {
+                for (int l = 0; l < QuadPtsPhy.size(); l++)
+                {
+                QuadWeights[l] = 0.866025403784439 * QuadWeights[l];
+                    for (int q = 0; q < dim; q++)
+                    {
+                    QuadPtsMapped[l][q] = 0.866025403784439 * QuadPtsMapped[l][q];
```

Comment:
If I understand correctly, this constant is the determinant of the linear map between reference element and current element? Should we not store this value e.g. in FEDDUtils or Helpers?